src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,98 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    public static final int DEFAULT_CAPACITY = 10;
+    public static final double CAPACITY_MULTIPLIER = 1.5;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkSize();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index of list");
+        }
+        checkSize();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
 
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValid(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexValid(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValid(index);
+        T removedElement = elements[index];
+        if (index + 1 != size) {
+            System.arraycopy(elements, index + 1, elements, index, size - index);
+        }
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == elements[i] || (elements[i] != null && elements[i].equals(element))) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("No such element exeption");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkSize() {
+        if (size == elements.length) {
+            T[] copy = (T[]) new Object[(int)(elements.length * CAPACITY_MULTIPLIER)];
+            System.arraycopy(elements, 0, copy, 0, size);
+            elements = copy;
+        }
     }
+
+    private void indexValid(int index) {
```

Comment:
`indexValid` - sounds like a statement.  `isIndexValid` - it is a more appropriate name for the validation method